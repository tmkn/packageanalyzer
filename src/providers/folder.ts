import * as path from "path";
import * as fs from "fs";
import * as semver from "semver";

import type { IPackageJson, IPackageMetadata, IUnpublishedPackageMetadata } from "../npm.js";
import { AbstractPackageProvider, type IPackageJsonProvider } from "./provider.js";

// todo replace in unit tests with FolderPackageProvider
// aka make FolderPackageProvider the default for unit tests

//gathers packages from a node_modules folder
export class FileSystemPackageProvider implements IPackageJsonProvider {
    private readonly _paths: Set<string> = new Set();
    private readonly _cache: Map<string, Map<string, IPackageJson>> = new Map();

    constructor(_folder: string) {
        const matches = this._findPackageJson(_folder);

        this._paths = new Set([...matches].sort((a, b) => a.localeCompare(b)));
        this._load();
    }

    private _findPackageJson(folder: string): string[] {
        try {
            const pkgs: string[] = [];

            const folderEntries = fs
                .readdirSync(folder, "utf8")
                .map(entry => path.join(folder, entry));

            for (const entry of folderEntries) {
                if (fs.statSync(entry).isDirectory()) {
                    pkgs.push(...this._findPackageJson(entry));
                } else if (entry.endsWith(`package.json`)) {
                    pkgs.push(entry);
                }
            }

            return pkgs;
        } catch (e) {
            console.log(e);

            return [];
        }
    }

    private _load(): void {
        const failedPaths: string[] = [];

        for (const pkgPath of this._paths) {
            try {
                const content = fs.readFileSync(pkgPath, "utf8");
                const pkg: IPackageJson = JSON.parse(content);

                this.addPackage(pkg);
            } catch {
                failedPaths.push(pkgPath);

                continue;
            }
        }

        if (failedPaths.length > 0) {
            console.log(`Failed to load ${failedPaths.length} packages`);
        }
    }

    addPackage(pkg: IPackageJson): void {
        const { name, version } = pkg;
        const versions = this._cache.get(name);

        if (typeof versions === "undefined") {
            this._cache.set(name, new Map([[version, pkg]]));
        } else {
            const specificVersion = versions.get(version);

            if (typeof specificVersion === "undefined") {
                versions.set(version, pkg);
            }
        }
    }

    async getPackageJson(name: string, version?: string | undefined): Promise<IPackageJson> {
        const versions = this._cache.get(name);
        let pkg: IPackageJson;

        if (typeof versions === "undefined") {
            throw new Error(`Couldn't find package ${name}`);
        }

        //load latest available version
        if (typeof version === "undefined") {
            const [latestVersion] = [...versions.keys()].slice(-1);

            if (!latestVersion) throw new Error(`Couldn't get latest version for ${name}`);

            const specificVersion = versions.get(latestVersion);

            //should never happen..
            if (typeof specificVersion === "undefined")
                throw new Error(`Error extracting latest package ${name}@${version}`);

            pkg = specificVersion;
        } else {
            const availableVersions: string[] = [...versions.keys()];
            const resolvedVersion = semver.maxSatisfying(availableVersions, version);

            if (resolvedVersion === null) {
                throw new Error(`Couldn't resolve ${version} for ${name}`);
            }

            const specificVersion = versions.get(resolvedVersion);

            if (typeof specificVersion === "undefined")
                throw new Error(`Couldn't find package ${name}@${version}`);

            pkg = specificVersion;
        }

        return pkg;
    }
}

// provides data provided by the dependencydump cli command
export class DumpPackageProvider extends AbstractPackageProvider {
    constructor(private readonly _folder: string) {
        super();

        this._loadData();
    }

    private _findPackageJson(folder: string): Set<string> {
        try {
            const pkgs: string[] = [];

            const folderEntries = fs
                .readdirSync(folder, "utf8")
                .map(entry => path.join(folder, entry));

            for (const entry of folderEntries) {
                if (fs.statSync(entry).isDirectory()) {
                    pkgs.push(...this._findPackageJson(entry));
                } else if (entry.endsWith(`metadata.json`)) {
                    pkgs.push(entry);
                }
            }

            return new Set([...pkgs].sort((a, b) => a.localeCompare(b)));
        } catch (e) {
            console.log(e);

            return new Set();
        }
    }

    private _loadData(): void {
        const paths = this._findPackageJson(this._folder);
        const failedPaths: string[] = [];

        for (const pkgPath of paths) {
            try {
                const content = fs.readFileSync(pkgPath, "utf8");
                const pkg: IPackageMetadata = JSON.parse(content);

                this._cache.set(pkg.name, pkg);
            } catch {
                failedPaths.push(pkgPath);

                continue;
            }
        }

        if (failedPaths.length > 0) {
            console.log(`Failed to load ${failedPaths.length} packages`);
        }
    }

    async getPackageMetadata(
        name: string
    ): Promise<IPackageMetadata | IUnpublishedPackageMetadata | undefined> {
        return this._cache.get(name);
    }
}
