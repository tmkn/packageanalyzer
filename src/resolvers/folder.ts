import * as path from "path";
import * as fs from "fs";

import { INpmPackageVersion } from "../npm";
import { EntryPackage } from "./resolver";

function getVersionFromPackageJson(rootPath: string): [string, string] {
    const packageJsonPath = path.join(rootPath, `package.json`);

    try {
        const content = fs.readFileSync(packageJsonPath, "utf8");
        const pkg: INpmPackageVersion = JSON.parse(content);

        return [pkg.name, pkg.version];
    } catch (e) {
        throw new Error(`Couldn't find package.json in ${rootPath}`);
    }
}

export function fromFolder(rootPath: string): EntryPackage {
    return getVersionFromPackageJson.bind(null, rootPath);
}
