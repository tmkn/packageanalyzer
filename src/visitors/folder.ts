import * as path from "path";
import * as fs from "fs";

import { INpmPackageVersion } from "../npm";
import { PackageVersion } from "./visitor";

export function getPackageJson(folder: string): PackageVersion {
    const packageJsonPath = path.join(folder, `package.json`);

    try {
        const content = fs.readFileSync(packageJsonPath, "utf8");
        const pkg: INpmPackageVersion = JSON.parse(content);

        return [pkg.name, pkg.version];
    } catch (e) {
        throw new Error(`Couldn't find package.json in ${folder}`);
    }
}
