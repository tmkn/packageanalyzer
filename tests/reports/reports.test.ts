import * as path from "path";

import { FileSystemPackageProvider } from "../../src/providers/folder.js";
import {
    createReportServiceFactory,
    type ITestReportNoValidationParams,
    TestReport,
    TestReportNoValidation
} from "../common.js";

describe(`ReportService Tests`, () => {
    const rootPath = path.join("tests", "data", "testproject1");
    const provider = new FileSystemPackageProvider(rootPath);

    const buildTestReport = createReportServiceFactory(TestReport, provider);

    test(`Executes report method`, async () => {
        const cb = vi.fn();
        const { reportService } = buildTestReport({
            pkg: [`react`],
            report: async () => {
                cb();
            }
        });

        await reportService.process();
        expect(cb).toHaveBeenCalledTimes(1);
    });

    test(`Executes multiple reports`, async () => {
        const cb = vi.fn();
        const params: ConstructorParameters<typeof TestReport> = [
            {
                pkg: [`react`],
                report: async () => {
                    cb();
                }
            }
        ];
        const { reportService } = buildTestReport(params, params, params);

        await reportService.process();
        expect(cb).toHaveBeenCalledTimes(3);
    });

    test(`Provides pkg argument`, async () => {
        let fullName: string = `Unknown`;
        const { reportService } = buildTestReport({
            pkg: [`react`],
            report: async ([pkg], _context) => {
                fullName = pkg.fullName;
            }
        });

        await reportService.process();

        expect(fullName).toEqual(`react@16.8.6`);
    });

    test(`Provides formatter argument`, async () => {
        const token = `Hello World`;
        const { reportService, stdout } = buildTestReport({
            pkg: [`react`],
            report: async (_, context) => {
                context.stdoutFormatter.writeLine(token);
            }
        });

        await reportService.process();

        expect(stdout.lines.find(line => line === token)).toBeDefined();
    });

    test(`Acknowledges depth setting`, async () => {
        let directDependenciesCount1: number = -1;
        let directDependenciesCount2: number = -1;

        const {
            reportService,
            reports: [report1, report2]
        } = buildTestReport(
            [
                {
                    pkg: [`react`],
                    report: async ([pkg], _context) => {
                        directDependenciesCount1 = pkg.directDependencies.length;
                    }
                }
            ],
            [
                {
                    pkg: [`react`],
                    report: async ([pkg], _context) => {
                        directDependenciesCount2 = pkg.directDependencies.length;
                    }
                }
            ]
        );

        report1.configs.depth = 0;
        report2.configs.depth = Infinity;

        await reportService.process();

        expect(directDependenciesCount1).toEqual(0);
        expect(directDependenciesCount2).toEqual(4);
    });

    test(`Writes to stderr on throw`, async () => {
        const { reportService, stdout, stderr } = buildTestReport({
            pkg: [`react`],
            report: async () => {
                throw new Error(`Whoopsie`);
            }
        });

        await reportService.process();

        expect(stdout.lines).toMatchSnapshot(`stdout`);
        expect(stderr.lines).toMatchSnapshot(`stderr`);
    });

    test(`Assigns data on missing validate method`, async () => {
        const params: ITestReportNoValidationParams = {
            foo: `blabla`
        };
        const report = new TestReportNoValidation(params);

        expect(JSON.stringify(params, null, 4)).toMatch(JSON.stringify(report.params, null, 4));
    });
});
