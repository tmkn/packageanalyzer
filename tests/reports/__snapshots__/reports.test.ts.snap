// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportService Tests Writes to stderr on throw 1`] = `
Array [
  "Error: Whoopsie",
  "",
]
`;
